<a href="https://colab.research.google.com/github/gonulsinem/1_Python_ML/blob/main/8_RegresyonFull.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

BTK Akademi- Python ile Makine Öğrenmesi Kursu calismalari ve kendi tuttuğum notlardan olusmaktadir.

Kursta verilen maaslar_yeni.csv veri seti kullanildi

In [1]:
#kutuphaneler import edildi
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
data = pd.read_csv('/content/maaslar_yeni.csv')
data

Calisan ID             unvan  UnvanSeviyesi  Kidem  Puan   maas
0            1             Cayci              1      5    70   2250
1            2          Sekreter              2      5    70   2500
2            3  Uzman Yardimcisi              3      5    70   3000
3            4             Uzman              4      5    70   4000
4            5  Proje Yoneticisi              5      5    70   5500
5            6               Sef              6      5    70   7500
6            7             Mudur              7      5    70  10000
7            8          Direktor              8      5    70  15000
8            9           C-level              9      5    70  25000
9           10               CEO             10      5    70  50000
10          11             Cayci              1      7    99   2000
11          12          Sekreter              2      7     9   2500
12          13  Uzman Yardimcisi              3      7    62   4000
13          14             Uzman              4      4    38   3000
14          15  Proje Yoneticisi              5      1    80   5000
15          16               Sef              6      2    35   5000
16          17             Mudur              7      8    99  12000
17          18          Direktor              8      4    58  11000
18          19           C-level              9      2    20  15000
19          20               CEO             10      4    42  22000
20          21             Cayci              1      8    11   2200
21          22          Sekreter              2      2    53   2200
22          23  Uzman Yardimcisi              3      8    50   2800
23          24             Uzman              4      9    91   6000
24          25  Proje Yoneticisi              5      9    71   5400
25          26               Sef              6      1     2   4000
26          27             Mudur              7     10    81  12000
27          28          Direktor              8     10    38  10000
28          29           C-level              9      1    50  15000
29          30               CEO             10      9    83  60000

In [4]:
data.head()
#maas bagimli degisken

Calisan ID             unvan  UnvanSeviyesi  Kidem  Puan  maas
0           1             Cayci              1      5    70  2250
1           2          Sekreter              2      5    70  2500
2           3  Uzman Yardimcisi              3      5    70  3000
3           4             Uzman              4      5    70  4000
4           5  Proje Yoneticisi              5      5    70  5500

In [5]:
data.describe()

Calisan ID  UnvanSeviyesi      Kidem       Puan          maas
count   30.000000      30.000000  30.000000  30.000000     30.000000
mean    15.500000       5.500000   5.433333  59.066667  10861.666667
std      8.803408       2.921384   2.686920  25.330823  13468.313281
min      1.000000       1.000000   1.000000   2.000000   2000.000000
25%      8.250000       3.000000   4.000000  44.000000   3000.000000
50%     15.500000       5.500000   5.000000  70.000000   5450.000000
75%     22.750000       8.000000   7.750000  70.000000  12000.000000
max     30.000000      10.000000  10.000000  99.000000  60000.000000

In [15]:
x = data.iloc[:,2:5]
X = x.values
x

UnvanSeviyesi  Kidem  Puan
0               1      5    70
1               2      5    70
2               3      5    70
3               4      5    70
4               5      5    70
5               6      5    70
6               7      5    70
7               8      5    70
8               9      5    70
9              10      5    70
10              1      7    99
11              2      7     9
12              3      7    62
13              4      4    38
14              5      1    80
15              6      2    35
16              7      8    99
17              8      4    58
18              9      2    20
19             10      4    42
20              1      8    11
21              2      2    53
22              3      8    50
23              4      9    91
24              5      9    71
25              6      1     2
26              7     10    81
27              8     10    38
28              9      1    50
29             10      9    83

In [16]:
y = data.iloc[:,-1:]
Y = y.values
y

maas
0    2250
1    2500
2    3000
3    4000
4    5500
5    7500
6   10000
7   15000
8   25000
9   50000
10   2000
11   2500
12   4000
13   3000
14   5000
15   5000
16  12000
17  11000
18  15000
19  22000
20   2200
21   2200
22   2800
23   6000
24   5400
25   4000
26  12000
27  10000
28  15000
29  60000

In [19]:
import statsmodels.api as sm

/usr/local/lib/python3.7/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [18]:
# MLR

from sklearn.linear_model import LinearRegression

lin_reg = LinearRegression()
lin_reg.fit(X,Y)


LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [21]:
model = sm.OLS(lin_reg.predict(X), X)
print(model.fit().summary())

                                 OLS Regression Results                                
Dep. Variable:                      y   R-squared (uncentered):                   0.903
Model:                            OLS   Adj. R-squared (uncentered):              0.892
Method:                 Least Squares   F-statistic:                              83.89
Date:                Sun, 05 Sep 2021   Prob (F-statistic):                    8.38e-14
Time:                        15:34:54   Log-Likelihood:                         -295.74
No. Observations:                  30   AIC:                                      597.5
Df Residuals:                      27   BIC:                                      601.7
Df Model:                           3                                                  
Covariance Type:            nonrobust                                                  
                 coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------

In [25]:
from sklearn.preprocessing import PolynomialFeatures
pol_reg = PolynomialFeatures(degree = 4)
x_pol = pol_reg.fit_transform(X)

lin_reg2 = LinearRegression()
lin_reg2.fit(x_pol,Y)

model2 = sm.OLS(lin_reg2.predict(pol_reg.fit_transform(X)),X)
print(model2.fit().summary())

                                 OLS Regression Results                                
Dep. Variable:                      y   R-squared (uncentered):                   0.680
Model:                            OLS   Adj. R-squared (uncentered):              0.644
Method:                 Least Squares   F-statistic:                              19.11
Date:                Sun, 05 Sep 2021   Prob (F-statistic):                    7.49e-07
Time:                        15:41:23   Log-Likelihood:                         -317.93
No. Observations:                  30   AIC:                                      641.9
Df Residuals:                      27   BIC:                                      646.1
Df Model:                           3                                                  
Covariance Type:            nonrobust                                                  
                 coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------

In [28]:
# SVR

from sklearn.preprocessing import StandardScaler

sc1 = StandardScaler()
x_scaler = sc1.fit_transform(X)

sc2 = StandardScaler()
y_scaler = sc2.fit_transform(Y)

from sklearn.svm import SVR
sv_reg = SVR(kernel = 'rbf')
sv_reg.fit(x_scaler, y_scaler)

model3 = sm.OLS(sv_reg.predict(x_scaler),x_scaler)
print(model3.fit().summary())


                                 OLS Regression Results                                
Dep. Variable:                      y   R-squared (uncentered):                   0.782
Model:                            OLS   Adj. R-squared (uncentered):              0.758
Method:                 Least Squares   F-statistic:                              32.37
Date:                Sun, 05 Sep 2021   Prob (F-statistic):                    4.34e-09
Time:                        15:47:32   Log-Likelihood:                        -0.92453
No. Observations:                  30   AIC:                                      7.849
Df Residuals:                      27   BIC:                                      12.05
Df Model:                           3                                                  
Covariance Type:            nonrobust                                                  
                 coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------

/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [29]:
# Decision Tree 

from sklearn.tree import DecisionTreeRegressor

dt_reg = DecisionTreeRegressor(random_state =0)
dt_reg.fit(X,Y)

model4 = sm.OLS(dt_reg.predict(X),X)
print(model4.fit().summary())

                                 OLS Regression Results                                
Dep. Variable:                      y   R-squared (uncentered):                   0.679
Model:                            OLS   Adj. R-squared (uncentered):              0.644
Method:                 Least Squares   F-statistic:                              19.08
Date:                Sun, 05 Sep 2021   Prob (F-statistic):                    7.62e-07
Time:                        15:47:44   Log-Likelihood:                         -317.95
No. Observations:                  30   AIC:                                      641.9
Df Residuals:                      27   BIC:                                      646.1
Df Model:                           3                                                  
Covariance Type:            nonrobust                                                  
                 coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------

In [31]:
from sklearn.ensemble import RandomForestRegressor
rf_reg = RandomForestRegressor(n_estimators = 10, random_state=0)
rf_reg.fit(X, Y.ravel())

model5 = sm.OLS(dt_reg.predict(X),X)
print(model5.fit().summary())

                                 OLS Regression Results                                
Dep. Variable:                      y   R-squared (uncentered):                   0.679
Model:                            OLS   Adj. R-squared (uncentered):              0.644
Method:                 Least Squares   F-statistic:                              19.08
Date:                Sun, 05 Sep 2021   Prob (F-statistic):                    7.62e-07
Time:                        15:49:03   Log-Likelihood:                         -317.95
No. Observations:                  30   AIC:                                      641.9
Df Residuals:                      27   BIC:                                      646.1
Df Model:                           3                                                  
Covariance Type:            nonrobust                                                  
                 coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------

# Tek Parametreli

In [32]:
data.corr() 
# maas ve unvan seviyesi arasinda 0.72

Calisan ID  UnvanSeviyesi     Kidem      Puan      maas
Calisan ID       1.000000       0.331847  0.206278 -0.251278  0.226287
UnvanSeviyesi    0.331847       1.000000 -0.125200  0.034948  0.727036
Kidem            0.206278      -0.125200  1.000000  0.322796  0.117964
Puan            -0.251278       0.034948  0.322796  1.000000  0.201474
maas             0.226287       0.727036  0.117964  0.201474  1.000000

In [39]:
x = data.iloc[:,2:3]
X = x.values
x

UnvanSeviyesi
0               1
1               2
2               3
3               4
4               5
5               6
6               7
7               8
8               9
9              10
10              1
11              2
12              3
13              4
14              5
15              6
16              7
17              8
18              9
19             10
20              1
21              2
22              3
23              4
24              5
25              6
26              7
27              8
28              9
29             10

In [41]:
# MLR

from sklearn.linear_model import LinearRegression

lin_reg = LinearRegression()
lin_reg.fit(X,Y)
model6 = sm.OLS(lin_reg.predict(X),X)
print(model6.fit().summary())

                                 OLS Regression Results                                
Dep. Variable:                      y   R-squared (uncentered):                   0.942
Model:                            OLS   Adj. R-squared (uncentered):              0.940
Method:                 Least Squares   F-statistic:                              468.1
Date:                Sun, 05 Sep 2021   Prob (F-statistic):                    1.93e-19
Time:                        15:52:40   Log-Likelihood:                         -287.43
No. Observations:                  30   AIC:                                      576.9
Df Residuals:                      29   BIC:                                      578.3
Df Model:                           1                                                  
Covariance Type:            nonrobust                                                  
                 coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------

In [42]:
from sklearn.preprocessing import PolynomialFeatures
pol_reg = PolynomialFeatures(degree = 4)
x_pol = pol_reg.fit_transform(X)

lin_reg2 = LinearRegression()
lin_reg2.fit(x_pol,Y)

model7 = sm.OLS(lin_reg2.predict(pol_reg.fit_transform(X)),X)
print(model7.fit().summary())

                                 OLS Regression Results                                
Dep. Variable:                      y   R-squared (uncentered):                   0.759
Model:                            OLS   Adj. R-squared (uncentered):              0.751
Method:                 Least Squares   F-statistic:                              91.39
Date:                Sun, 05 Sep 2021   Prob (F-statistic):                    1.82e-10
Time:                        15:53:26   Log-Likelihood:                         -311.94
No. Observations:                  30   AIC:                                      625.9
Df Residuals:                      29   BIC:                                      627.3
Df Model:                           1                                                  
Covariance Type:            nonrobust                                                  
                 coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------

In [43]:
# SVR

from sklearn.preprocessing import StandardScaler

sc1 = StandardScaler()
x_scaler = sc1.fit_transform(X)

sc2 = StandardScaler()
y_scaler = sc2.fit_transform(Y)

from sklearn.svm import SVR
sv_reg = SVR(kernel = 'rbf')
sv_reg.fit(x_scaler, y_scaler)

model8 = sm.OLS(sv_reg.predict(x_scaler),x_scaler)
print(model8.fit().summary())

                                 OLS Regression Results                                
Dep. Variable:                      y   R-squared (uncentered):                   0.770
Model:                            OLS   Adj. R-squared (uncentered):              0.762
Method:                 Least Squares   F-statistic:                              97.19
Date:                Sun, 05 Sep 2021   Prob (F-statistic):                    9.11e-11
Time:                        15:54:08   Log-Likelihood:                         -1.3680
No. Observations:                  30   AIC:                                      4.736
Df Residuals:                      29   BIC:                                      6.137
Df Model:                           1                                                  
Covariance Type:            nonrobust                                                  
                 coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------

/usr/local/lib/python3.7/dist-packages/sklearn/utils/validation.py:760: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


In [44]:
# Decision Tree 

from sklearn.tree import DecisionTreeRegressor

dt_reg = DecisionTreeRegressor(random_state =0)
dt_reg.fit(X,Y)

model9 = sm.OLS(dt_reg.predict(X),X)
print(model9.fit().summary())

                                 OLS Regression Results                                
Dep. Variable:                      y   R-squared (uncentered):                   0.751
Model:                            OLS   Adj. R-squared (uncentered):              0.742
Method:                 Least Squares   F-statistic:                              87.29
Date:                Sun, 05 Sep 2021   Prob (F-statistic):                    3.01e-10
Time:                        15:54:50   Log-Likelihood:                         -312.62
No. Observations:                  30   AIC:                                      627.2
Df Residuals:                      29   BIC:                                      628.6
Df Model:                           1                                                  
Covariance Type:            nonrobust                                                  
                 coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------

In [45]:
from sklearn.ensemble import RandomForestRegressor
rf_reg = RandomForestRegressor(n_estimators = 10, random_state=0)
rf_reg.fit(X, Y.ravel())

model10 = sm.OLS(dt_reg.predict(X),X)
print(model10.fit().summary())

                                 OLS Regression Results                                
Dep. Variable:                      y   R-squared (uncentered):                   0.751
Model:                            OLS   Adj. R-squared (uncentered):              0.742
Method:                 Least Squares   F-statistic:                              87.29
Date:                Sun, 05 Sep 2021   Prob (F-statistic):                    3.01e-10
Time:                        15:55:43   Log-Likelihood:                         -312.62
No. Observations:                  30   AIC:                                      627.2
Df Residuals:                      29   BIC:                                      628.6
Df Model:                           1                                                  
Covariance Type:            nonrobust                                                  
                 coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------

 # Tek Parametreli
 1. R-squared (uncentered):                   0.942 Lin Reg
 2. R-squared (uncentered):                   0.759 Poly Reg
 3. R-squared (uncentered):                   0.770 SVR
 4. R-squared (uncentered):                   0.751 Decision Tree
 5. R-squared (uncentered):                   0.751 Random Forest

## 3 Parametreli
1. R-squared (uncentered):                   0.903 LinReg

2. R-squared (uncentered):                   0.680 PolyReg

3. R-squared (uncentered):                   0.782 SVR

4. R-squared (uncentered):                   0.679 Decision Tree

5.  R-squared (uncentered):                   0.679 Random Forest